Image support for Overlays

gfargo opened the pull request:
We now have the ability to feature images as an FM-Overlay.  All images added should scale to the respective device based on aspect ration & available screen real-estate.
### Features
- Overlays are displayed every 20hours using a cookie set inside of the main `class-fm-overlays.php`.
- Adds `class-fm-overlays-helpers.php` with a few template helper functions
- Overlays now feature dynamic classes, created in `get_overlay_classes()` within the helpers class
- Adds responsive picture element that is populated with all available image sizes
#### Notable
- Instead of making the ‘get_post_meta()`call inside of the the template, I added it to the post object inside of`class-fm-overlays.php` line 343
- added two new fields to what was there, one for the image link another ‘checkbox’ to have the link open in a new window (edited)
- Currently image sizes are using default [WP Thumbnail sizes](https://codex.wordpress.org/Post_Thumbnails) e.g. `medium`, `large`, and `full`.
##### Fields Added
- Image Field
- Image Link 
- Image Link Target

jakewrfoster committed:
adding image content

jakewrfoster committed:
image field and tempalte

Commit:
add link field

- updates fields on FM-Overlays to have a link field when image is selected as content type

Commit:
adds field for link target

Commit:
updates overlays class and templates to support overlay image partial

- adds meta content for overlay to $overlay post object in class
- adds helper class for templates and repsonsive image partial
- adds responsive image partial
- adds dynamic classes to fm-overlay windows

Commit:
frontend changes to fm-overlays image partial

Commit:
adds working javascript calculations for fm-overlays

Commit:
adds additional class to image partial

Commit:
Clean up JS

- adds a few variables
- adds a few comments to functions

Commit:
Loads Helper Class in Main Plugin File

removes helper class from `class-fm-overlays.php` and adds it to the main plugin file that is run at init.  Simplifies things a bit.

Commit:
removes outline glow when clicking close button

removes default browser glow around close button and replaces with reduced opacity.

Commit:
improves image resizing

fixes bug where on initial display image would be misformatted

Commit:
adds cookie support for overlays

- removes old unused code
- adds support for cookies on overlays / overlays will now only display every 20 hours to user

Commit:
fixes missing semicolon in js

Commit:
fixes phpcs / adds additional comments

Commit:
phpcs fixes

Commit:
updates javascript 

- moves variable declaration into main function scope
- changes const to let
- updates comments in code
- moves overlay image conditional into timeout function

Commit:
adds back 'get_conditionals' function to class

whoops! somehow this got deleted in a previous commit

Commit:
fixes weird height bug

sometimes at specific resolutions there would be a 1px high border below the image due to oddities in the calculation.  
- changes js to use math.ceil() on certain height variables
- adds margin-bottom: -1px to overlay image class

Commit:
switches image size to default 'medium'

Commit:
fixes phpcs issues

adds correct spacing to class

jakewrfoster commented on client/js/global.js:
```diff
@@ -17,26 +17,75 @@ const $ = require('jQuery');
   const $overlay = $('#fm-overlay');
   const $overlayWrapper = $overlay.children('.fm-overlay-wrapper');
+  const $overlayFade = $overlay.children('.fm-overlay-fade');
+  const $overlayImage = $overlay.find('img', '.fm-image');
```
Would this also grab an inlined image from a rich text editor? If so, we want to make sure we don't bork any inline styles those images may have received.

jakewrfoster commented on client/js/global.js:
```diff
@@ -17,26 +17,75 @@ const $ = require('jQuery');
+       */
+      if ($overlay.hasClass('fm-overlay-image')) {
+        // handle image overlay resizing
+        resizeOverlayImage();
```
regarding the comment above: gotcha

jakewrfoster commented on client/js/global.js:
```diff
@@ -17,26 +17,75 @@ const $ = require('jQuery');
+      /**
+       * Check & Handle Image Overlays
+       */
+      if ($overlay.hasClass('fm-overlay-image')) {
```
will the functions below be firing even if the overlay is not visible?

gfargo replied:
Ah good call.. Updated version of this removes the `resize` event listener inside of the `hideOverlay` function.

Commit:
optimizes javascript for overlay images

- triggers all image related resizing after overlay is visible
- targets only images inside overlay image types
- updates comments

gfargo commented on client/js/global.js:
```diff
@@ -21,7 +21,7 @@ function fmOverlay() {
   const $overlayWrapper = $overlay.children('.fm-overlay-wrapper');
   const $overlayFade = $overlay.children('.fm-overlay-fade');
-  const $overlayImage = $overlay.find('img', '.fm-image');
+  const $overlayImage = $overlay.find('img', '.fm-overlay-content.image');
```
Changing `.fm-image` to `.fm-overlay-content.image` will ensure that only images inside of the overlay-image will be targeted and all images inside of a rich text overlay will be left alone.

gfargo commented on client/js/global.js:
```diff
@@ -73,18 +77,18 @@ function fmOverlay() {
        */
-      if ($overlay.hasClass('fm-overlay-image')) {
-        // handle image overlay resizing
+      if ($overlay.show().addClass(activeClass).hasClass('fm-overlay-image')) {
```
Sequentially chains together the display of the modal to the resizing of the contained image.  This should prevent against race conditions where `resizeOerlayImage()` is fired before overlay is visible.

gfargo commented:
Pushed up a few changes that address some of the comments on the previous commit.  There shouldn't  be any conflict between functionality for Image overlays bleeding into Rich Text overlays.

zgreen commented on client/css/global.scss:
```diff
@@ -45,6 +51,13 @@ $overlay-index: 5000;
+		// removes default outline glow on click
+		&:focus,
+		&:active {
+			outline: none;
```
http://www.outlinenone.com/

zgreen commented on php/class-fm-overlays-helpers.php:
```diff
@@ -0,0 +1,79 @@
+			);
+
+			// set_transient( $transient_slug, $image_sizes );
+		// }
```
commented code

zgreen commented on php/class-fm-overlays-helpers.php:
```diff
@@ -0,0 +1,79 @@
+
+	public function get_overlay_image_sizes( $attachment_id ) {
+		$transient_slug = 'fm_overlay_image_sizes_' . $attachment_id;
+		// if ( false === ( $image_sizes = get_transient( $transient_slug ) ) ) {
```
commented code

gfargo replied:
good catch! I added some `@todo` to remind use we probably eventually want to add caching to this function.

zgreen commented on php/class-fm-overlays.php:
```diff
@@ -181,18 +180,44 @@ public function destroy_transient( $post_id ) {
+		/**
+		 * Cookie will expire in 20 Hours
+		 */
+		setcookie( $cookie_name, true, time() + 60 * 60 * 20 );
```
It may be worth noting this in the docs somewhere, since EU laws require websites to notify users that a website sets cookies.

jakewrfoster replied:
Yeah, I will add something in the README about this. Ultimately, this should fall to the individual site to address, not the plugin. The plugin _should_ reiterate this in the README.

gfargo replied:
I added a short section in the main Readme with this updated commit @romancandlethoughts, but feel free to update however you want.

zgreen commented on templates/fm-overlay-basic.php:
```diff
@@ -24,12 +27,19 @@
+						target="<?php echo ( ! empty( $overlay->overlay_content['image_link_target'] ) ? '_blank' : '' ); ?>"
+						alt="<?php echo esc_attr( $overlay->post_title ); ?>"
+						class="fm-image-link">
+						<?php include( FM_OVERLAYS_PATH . 'templates/partials/partial-overlay-image.php' ); ?>
```
Why not `get_template_part()` here?

gfargo replied:
I believe `get_template_part()` is only used within Themes [because it relies](http://wordpress.stackexchange.com/questions/124789/how-to-include-a-file-using-get-template-part-in-a-plugin/124794#124794) on the `locate_template` [function](https://codex.wordpress.org/Function_Reference/locate_template).   I could be misinterpreting this   though..

zgreen replied:
right, sorry

zgreen commented on templates/partials/partial-overlay-image.php:
```diff
@@ -0,0 +1,47 @@
+
+
+<?php if ( ! empty( $image_sizes ) ) : ?>
+	<picture <?php if ( ! empty( $id ) ) : ?>
```
Many browsers still don't support `<picture>` (http://caniuse.com/#feat=picture). Does this plugin use [picturefill](https://github.com/scottjehl/picturefill)? (If not, it should.)

zgreen replied:
Also, it looks like there's no `$id`.

gfargo replied:
Broke out adding [picturefill](https://github.com/scottjehl/picturefill) into its own issue; #12.

zgreen commented:
A few notes/questions. I can review it again once those are addressed, thanks!

Commit:
adds default outline back to close link

Commit:
removes commented out code and adds addtional documentation

Commit:
updates readme with notices about using cookies

jakewrfoster commented on README.md:
```diff
@@ -5,8 +5,16 @@ A Fieldmanager extension to render dynamic content in modal boxes/overlays on a
+
+Cookies are set in the `set_overlay_cookie` function inside of `class-fm-overlays.php` using the overlay ID to dynamically generate a name like `fm-overlay-2357`.
+
+*Important:* When implementing on a site visitors *must be notified* cookies are being set [per EU Guidelines](http://ec.europa.eu/ipg/basics/legal/cookies/index_en.htm).
```
_Important:_ When implementing on a site visitors **must be notified** cookies are being set [per EU Guidelines](http://ec.europa.eu/ipg/basics/legal/cookies/index_en.htm). This plugin does not handle any such notifications.

gfargo replied:
Got it.  Thanks man

zgreen commented:
🚀

Commit:
updates readme for cookies
